### PR TITLE
Improved Performance Logging

### LINKED ISSUE #8345: Improve Performance Logging
### Description of the enhancement or error report

Our performance logging is still lacking in several ways.

### Rationale for the enhancement or information for reproducing the error

Specifically, I noticed today that we have no timers for UOs (including PPs and VPPs with `execute_on = LINEAR NONLINEAR`. This is important for one of the PRs @YaqiWang put in for an optimization. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

No impacts, just better logging.

### PR DESCRIPTION
**Pull request by permcody**
This change makes the "Execution" section of the performance log much more dynamic.
Only active objects are logged and as more objects are added to a simulation, more
lines appear in the log to categorize them.

More importantly, this also now logs LINEAR and NONLINEAR UOs which was completely
missing before.

closes #8345

Ugh - This PR "looks" huge but it's not. There are several indentation changes because of the check to see if there are active objects in each section prior to entering that section and starting logs.

### CODE REVIEW

**Comment by YaqiWang** [comment]
I always hope the differ is smarter to figure out the indentations...

**Comment by permcody** [comment]
See https://github.com/idaholab/moose/issues/8345 for sample output using this new logic.

**Comment by permcody** [comment]
No big rush on this PR, I think it continues to improve the performance logging by making the Execution output section only print parts that are actually executed. @idaholab/moose-team can take a look and vote.

**Review by andrsd — APPROVED** [review]

File: framework/src/base/AuxiliarySystem.C [hunk 1]
```diff
@@ -396,14 +405,17 @@ AuxiliarySystem::computeNodalVars(ExecFlagType type)
...
+    Moose::perf_log.pop(compute_aux_tag, "Execution");
   }
-  PARALLEL_CATCH;
-  Moose::perf_log.pop("update_aux_vars_nodal()", "Execution");
 
-  // Boundary Nodal AuxKernels
-  Moose::perf_log.push("update_aux_vars_nodal_bcs()", "Execution");
-  PARALLEL_TRY {
-    if (nodal.hasActiveBoundaryObjects())
+  if (nodal.hasActiveBoundaryObjects())
```
Comment (jwpeterson) [inline framework/src/base/AuxiliarySystem.C]: Are we sure that all processors always return the same result for `if (nodal.hasActiveBoundaryObjects())` et al?  If not, moving the `PARALLEL_CATCH` inside the if-statement will cause the code to hang.

Comment (friedmud) [inline framework/src/base/AuxiliarySystem.C]: @jwpeterson I think that they _do_... but we should double check.

**Review by friedmud — APPROVED** [review]
This looks good to me... but double check on what @jwpeterson asked about the parallel consistency of `hasActive*`...

File: framework/src/base/AuxiliarySystem.C (hunk 1 above)
Comment (aeslaughter) [inline framework/src/base/AuxiliarySystem.C]: I can't image that they are any different across processors, the list of boundary/block stuff in the warehouses is populated from the 'boundary/block' input parameters. If those are not present then all blocks/boundaries are used from the mesh object.

I guess in that the complete list of blocks could be incomplete on a distributed mesh.

Comment (permcody) [inline framework/src/base/AuxiliarySystem.C]: I've been looking at the warehouses and am convinced that the various `Warehouse::has...Objects()` methods are all symmetric in parallel runs. However, I have a similar check for a StoredRange elsewhere in this PR that I'm pretty sure isn't symmetric. I'd probably better undo that one...

File: framework/src/base/NonlinearSystemBase.C [hunk 2]
```diff
@@ -1143,27 +1162,35 @@ NonlinearSystemBase::computeNodalBCs(NumericVector<Number> & residual)
...
     _fe_problem.getAuxiliarySystem().solution().close();
 
-  PARALLEL_TRY {
-    // last thing to do are nodal BCs
+  PARALLEL_TRY
+  {
     ConstBndNodeRange & bnd_nodes = *_mesh.getBoundaryNodeRange();
-    for (const auto & bnode : bnd_nodes)
+
+    if (!bnd_nodes.empty())
```
Comment (permcody) [inline framework/src/base/NonlinearSystemBase.C]: I think this condition could diverge in a parallel run.

Comment (permcody) [inline framework/src/base/NonlinearSystemBase.C]: Ok - not parallel runs but DistributedMesh but that's another story

Comment (friedmud) [inline framework/src/base/NonlinearSystemBase.C]: Good catch!

**Comment by permcody** [comment]
It looks like bnd_nodes is already broken for DistributedMesh even now. We
populate it with MeshBase::node() which will fail badly in that case. I'll
open another ticket to fix that. In general, we need to go to through
framework and remove all of those types of calls.

**Comment by friedmud** [comment]
Ok - but will this PR cause DistributedMesh runs to hang?

**Comment by permcody** [comment]
No, in fact they "run" just fine as long as assertions are turned off even
thought he behavior is undefined. I think the reason we don't crash is
because we rarely access those off processor elements or nodes so we get
away with it.

We have several places in the framework were we just access nodes and
elements without using the pointer versions - we really need to clean them
all up. I'll start with this one and go from there.

**Comment by friedmud** [comment]
I still don't understand your answer.  The question is: can this be merged - or is it causing some extra breakage?

**Comment by permcody** [comment]
Let's not merge just yet. I'll fix the DistributedMesh issue first and add
that as a commit.

File: framework/src/base/NonlinearSystemBase.C (hunk 2 above)
Comment (permcody) [inline framework/src/base/NonlinearSystemBase.C]: OK - so I dug into this quite a bit more and learned several things. First, the boundary node range, as managed by `MooseMesh` works fine on DistributedMesh. We use the BoundaryInfo object in libMesh to get this information and it works properly only returning parts of the mesh that are accessible on each processor. From the docs:
``` 
 /**
   * Creates a list of nodes and ids for those nodes.
   *
   * On a ReplicatedMesh this will include all nodes; on a
   * DistributedMesh only semilocal nodes will be included.
   */
```

So, on a DistributedMesh, some processors may enter this branch, while others might not. I have verified that none of the contents of this branch contain and parallel communication.

**Comment by permcody** [comment]
This is ready to be reviewed. The loop I thought was problematic (`if (!bnd_nodes.empty())`) is fine with DistributedMesh.
